### DISCUSSION TITLE
Changing material for glb at runtime.

### DISCUSSION THREAD

**Discussion by atul161** [google/filament#6431] [General]
Hi @romainguy  and the team,

We are going good with the filament library and it is performing well in most of the device.
I have some question and need your help with that.

I have a vrm model 

- Can i change the material of glb at runtime?
- Suppose i have an avatar and i want to control its skin color , darkness and want to put some asset to. ie: Add cap on top of 
avatar or change the clothes of avatar.Just want to know that how filament is useful for all these kind of operations.

**Comment by romainguy** [comment]
You can change materials and/or material properties at runtime, yes. You can see the names of the parameters in the glTF material definitions: https://github.com/google/filament/blob/main/libs/gltfio/materials/base.mat.in

For instance to change skin color, you'd change the `baseColorFactor` or `baseColorMap` properties. You could also use glTF's [material variants](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_variants/README.md) feature, which is supported by Filament.

**Comment by Bolloxim** [comment]
Hi Romain,
I'd like to follow up to this original question as I am finding a fundamental problem in using a custom shader for an effect on a model (2side1pass transparent).
If I create a MaterialInstance using a new material instance with a modified shader that uses the baseColorMap sampler and then switch out the material instance.  
The texture does not get bound.  I cannot read the original texture from the GLB file so cannot manually bind it.  I had hoped it would just bind to existing samplers, However reviewing the code in gltfio the material instance generates all the bindings from the source material and caches material instances, so basically its already bound therefore it doesn't actually do any binding..  very efficient from a prebuilt perspective. 

Is there a method to be able to swap out a GLB material instance with a custom material but have the parameters bind. I have had idea's of generating new shader variants for the ubershader and then process the glb to have material variants, this is cumbersome though because each new model effect will have to have a new variant and built.  I looked at just adding dynamic variants on loading but the glb is basically immutable making it impossible without rewriting code to dynamically add variants at runtime.

I would appreciate suggestions on how we can achieve material effects on a GLB without tooling each one in as baked variants

-andi

**Comment by nmfisher** [comment]
Loosely related to the original question - is there any way to set an Ubershader material based on the label/index that was used to build it? e.g. in `libs/gltfio/CMakeLists.txt`, we have:

```
# NOTE: The default ubershader material is defined as the first material in the list below.

build_ubershader(lit_opaque                base            lit opaque)
build_ubershader(lit_fade                  base            lit fade)
build_ubershader(lit_masked                base            lit masked)
build_ubershader(specularGlossiness_fade   base            specularGlossiness fade)
build_ubershader(specularGlossiness_opaque base            specularGlossiness opaque)
build_ubershader(specularGlossiness_masked base            specularGlossiness masked)
build_ubershader(unlit_fade                base            unlit fade)
build_ubershader(unlit_opaque              base            unlit opaque)
build_ubershader(unlit_masked              base            unlit masked)
build_ubershader(volume                    volume          _ _)
build_ubershader(transmission              transmission    _ _)
build_ubershader(sheen                     sheen           _ _)
```

It would be convenient to use a single Ubershader provider to load one asset with `lit_fade` (or 1), then load some other asset with `unlit_opaque` (or 7) (basically ignoring the properties of the glTF asset and forcing a particular ubershader material).
